include/nbl/builtin/glsl/blit/blit/blit.glsl

```diff
@@ -0,0 +1,177 @@
+nbl_glsl_blit_parameters_t nbl_glsl_blit_getParameters();
+
+vec4 nbl_glsl_blit_getData(in vec3 texCoord, in uint layerIdx);
+void nbl_glsl_blit_setData(in vec4 data, in uvec3 coord, in uint layerIdx);
+vec4 nbl_glsl_blit_getKernelWeight(in uint index);
+void nbl_glsl_blit_addToHistogram(in uint bucketIndex, in uint layerIdx);
+
+#ifndef _NBL_GLSL_BLIT_ALPHA_BIN_COUNT_
+#error _NBL_GLSL_BLIT_ALPHA_BIN_COUNT_ must be defined
+#endif
+
+#define scratchShared _NBL_GLSL_SCRATCH_SHARED_DEFINED_
+
+uvec3 linearIndexTo3DIndex(in uint linearIndex, in uvec3 gridDim)
+{
+	uvec3 index3D;
+
+	const uint areaXY = gridDim.x * gridDim.y;
+	index3D.z = linearIndex / areaXY;
+	index3D.y = (linearIndex - (index3D.z * areaXY)) / gridDim.x;
+	index3D.x = linearIndex - (index3D.y*gridDim.x) - (index3D.z*areaXY);
+
+	return index3D;	
+}
+
+uint nbl_glsl_multi_dimensional_array_addressing_snakeCurve(in uvec3 coordinate, in uvec3 extents)
+{
+	return (coordinate.z * extents.y + coordinate.y) * extents.x + coordinate.x;
+}
+
+// p in input space
+ivec3 getMinKernelWindowCoord(in vec3 p, in vec3 minSupport)
+{
+	return ivec3(ceil(p - vec3(0.5f) + minSupport));
+}
+
+void nbl_glsl_blit_main()
+{
+	const nbl_glsl_blit_parameters_t params = nbl_glsl_blit_getParameters();
+
+	const uvec3 outputTexelsPerWG = params.outputTexelsPerWG;
+
+	const vec3 scale = params.fScale;
+	const vec3 halfScale = scale * vec3(0.5f);
+
+	const uvec3 minOutputPixel = gl_WorkGroupID * outputTexelsPerWG;
+	const vec3 minOutputPixelCenterOfWG = vec3(minOutputPixel)*scale + halfScale;
+	const ivec3 regionStartCoord = getMinKernelWindowCoord(minOutputPixelCenterOfWG, params.negativeSupport); // this can be negative, in which case we wrap
+
+	const uvec3 preloadRegion = params.preloadRegion;
+
+	for (uint virtualInvocation = gl_LocalInvocationIndex; virtualInvocation < preloadRegion.x * preloadRegion.y * preloadRegion.z; virtualInvocation += _NBL_GLSL_WORKGROUP_SIZE_)
+	{
+		const ivec3 inputPixelCoord = regionStartCoord + ivec3(linearIndexTo3DIndex(virtualInvocation, preloadRegion));
+
+		vec3 inputTexCoord = (inputPixelCoord + vec3(0.5f)) / params.inDim;
+		
+		const vec4 loadedData = nbl_glsl_blit_getData(inputTexCoord, gl_WorkGroupID.z);
+		for (uint ch = 0; ch < _NBL_GLSL_BLIT_OUT_CHANNEL_COUNT_; ++ch)
+			scratchShared[ch][virtualInvocation] = loadedData[ch];
+	}
+	barrier();
+
+	const uvec3 iterationRegions[3] = uvec3[]( uvec3(outputTexelsPerWG.x, preloadRegion.yz), uvec3(outputTexelsPerWG.yx, preloadRegion.z), outputTexelsPerWG.yxz );
+
+	uint read = 0;
+	for (uint axis = 0; axis < _NBL_GLSL_BLIT_DIM_COUNT_; ++axis)
+	{
+		const uvec3 iterationRegion = iterationRegions[axis];
+		for (uint virtualInvocation = gl_LocalInvocationIndex; virtualInvocation < iterationRegion.x * iterationRegion.y * iterationRegion.z; virtualInvocation += _NBL_GLSL_WORKGROUP_SIZE_)
+		{
+			const uvec3 virtualInvocationID = linearIndexTo3DIndex(virtualInvocation, iterationRegion);
```

Comment:
after the compiler unrolls the loop `const uvec3 iterationRegion` becomes free (it will just constant fold), so what you did is okay

however `iterationRegion.x * iterationRegion.y * iterationRegion.z` won't be, I'd supply that though a `uvec3` (one component per axis) push constant, 

The `iterationRegion.x * iterationRegion.y` as well, so you could use `nbl_glsl_multi_dimensional_array_addressing_snakeCurveInverse (in uint linearIndex, in uvec2 gridDimPrefixProduct)` instead of `linearIndexTo3DIndex`

Reply:
I don't get it, I thought we [sacrificed loop unrolling for less pipeline combos and ability to cache them](https://github.com/Devsh-Graphics-Programming/Nabla/pull/343#discussion_r900007185) ?

Do you want me to:
1. Have 3 separate `uvec3` push constants each with `iterationRegions[i]`? Or,
2. Have 1 `uvec3` push constant with their products `{iterationRegions[0].x*iterationRegions[0].y*iterationRegions[0].z, .., ..}`
?

Reply:
> The iterationRegion.x * iterationRegion.y as well, so you could use nbl_glsl_multi_dimensional_array_addressing_snakeCurveInverse (in uint linearIndex, in uvec2 gridDimPrefixProduct) instead of linearIndexTo3DIndex

Does this mean another `uvec3` in the push constants, like:
`{iterationRegions[0].x*iterationRegions[0].y, .., ..}`
?

Reply:
Inclusive prefix products?

`{iterationRegions[0].x, iterationRegions[0].x*iterationRegions[0].y, iterationRegions[0].x*iterationRegions[0].y*iterationRegions[0].z}`

and, similarly for `iterationRegions[1]` and `iterationRegions[2]`

Reply:
I do think this is what you meant because, `iterationRegions[i].x` and `iterationRegions[i].x*iterationRegions[i].y` are directly usable for `nbl_glsl_multi_dimensional_array_addressing_snakeCurveInverse (in uint linearIndex, in uvec2 gridDimPrefixProduct)` and `iterationRegions[i].x*iterationRegions[i].y*iterationRegions[i].z` is directly usable as the upper limit of the `virtualInvocation` `for` loop, but I still want to confirm.

Reply:
> Inclusive prefix products?
> 
> `{iterationRegions[0].x, iterationRegions[0].x*iterationRegions[0].y, iterationRegions[0].x*iterationRegions[0].y*iterationRegions[0].z}`
> 
> and, similarly for `iterationRegions[1]` and `iterationRegions[2]`

inclusive yes, thats why I wanted a `snakeCurve` addressing util function that takes a `uvec2`

and yes the last one is usable as the upper limit for `virtualInvocation`

Reply:
Resolved.